### PR TITLE
Add debugging capabilities to features.py

### PR DESCRIPTION
**Pull request by m09**
Small PR to allow for debugging of features.py. It's what I've been using to browse the different errors we get during parsing.

Signed-off-by: Hugo Mougard <hugo@sourced.tech>

### CODE REVIEW

**Comment by m09** [comment]
This should be merged after the next one that I push btw, not now.

File: lookout/style/format/features.py [hunk 1]
```diff
@@ -43,6 +43,7 @@ def __init__(self,  value: str, start: Position, end: Position,
         self.value = value
         assert start.line >= 1 and start.col >= 1, "start line and column are 1-based like UASTs"
         assert end.line >= 1 and end.col >= 1, "end line and column are 1-based like UASTs"
+        assert y in EMPTY_CLS or start.offset < end.offset, "illegal empty node"
```
Comment (EgorBu) [inline lookout/style/format/features.py]: Could you clarify this line? Without context, I can't understand why it's required

File: lookout/style/format/features.py [hunk 2]
```diff
@@ -83,6 +84,12 @@ def from_node(node: bblfsh.Node, file: str, path: str) -> Iterable["VirtualNode"
                               node=node, path=path)
             return
         start_offset = outer_token.find(node.token)
+        assert start_offset >= 0, (
+            "Couldn't find the token in the specified position:\nNode role: %s\nParsed form: “%s”"
```
Comment (EgorBu) [inline lookout/style/format/features.py]: File information is needed

File: lookout/style/format/features.py [hunk 3]
```diff
@@ -232,9 +243,12 @@ def extract_features(self, files: Iterable[File], lines: List[List[int]]=None
             try:
                 vnodes, parents = self._parse_file(contents, uast, file.path)
             except AssertionError as e:
-                self._log.warning("could not parse file %s with error \"%s\', skipping",
+                self._log.warning("could not parse file %s with error '%s', skipping",
                                   file.path, e)
-                continue
+                if self.debug_parsing:
```
Comment (EgorBu) [inline lookout/style/format/features.py]: Alternative is to use https://github.com/src-d/ml/blob/69361199e04f393d1a10b10badb969a3f5b2ed24/sourced/ml/utils/engine.py#L83-L92

File: lookout/style/format/features.py [hunk 4]
```diff
@@ -146,12 +153,15 @@ def to_comment(self, correct_y: int) -> Comment:
 CLASSES = (CLS_SPACE, CLS_TAB, CLS_NEWLINE, CLS_SPACE_INC, CLS_SPACE_DEC,
            CLS_TAB_INC, CLS_TAB_DEC, CLS_SINGLE_QUOTE, CLS_DOUBLE_QUOTE, CLS_NOOP)
 CLASS_INDEX = {cls: i for i, cls in enumerate(CLASSES)}
+EMPTY_CLS = frozenset([CLASS_INDEX[CLS_TAB_DEC], CLASS_INDEX[CLS_SPACE_DEC],
```
Comment (EgorBu) [inline lookout/style/format/features.py]: why do you need to extract hash value from `EMPTY_CLS`? Or you want to make this structure unmodifiable?

File: lookout/style/format/features.py [hunk 5]
```diff
@@ -232,9 +243,12 @@ def extract_features(self, files: Iterable[File], lines: List[List[int]]=None
             try:
                 vnodes, parents = self._parse_file(contents, uast, file.path)
             except AssertionError as e:
-                self._log.warning("could not parse file %s with error \"%s\', skipping",
+                self._log.warning("could not parse file %s with error '%s', skipping",
                                   file.path, e)
-                continue
+                if self.debug_parsing:
+                    input("press Enter to continue")
```
Comment (vmarkovtsev) [inline lookout/style/format/features.py]: How about `import ipdb; ipdb.set_trace()` then?

File: lookout/style/format/features.py (hunk 1 above)
Comment (m09) [reply lookout/style/format/features.py]: It's a test that empty nodes are only of the classes from which we expect empty nodes (-space, -tab and noop)

File: lookout/style/format/features.py (hunk 2 above)
Comment (m09) [reply lookout/style/format/features.py]: file information is already present where all the asserts are caught

File: lookout/style/format/features.py (hunk 3 above)
Comment (m09) [reply lookout/style/format/features.py]: true but I feel like it's more verbose, needs an import and doesn't filter on exception type so maybe it's better to stick to that

File: lookout/style/format/features.py (hunk 4 above)
Comment (m09) [reply lookout/style/format/features.py]: it's a reflex: when I use something primarily for `in` operations I use sets. And this won't be modified so `frozenset` in this case.

**Comment by m09** [comment]
This one should pass travis after #123 merge

File: lookout/style/format/features.py (hunk 5 above)
Comment (m09) [reply lookout/style/format/features.py]: I went for printing the trace directly instead, it proved to be the fatest during debugging. I can change if needed.

**Pull request merged**